source/base/DateTime.ooc

```diff
@@ -14,22 +14,251 @@
+
+	now: static DateTime {
+		get {
+					DateTime new(Time year(), Time month(), Time day(), Time hour(), Time min(), Time sec(), Time microsec()/1000)
+		}
+	}
+
+	_ticksToDateTimeHelper: static func (totalTicks: Int64) -> DateTimeData {
+		fourYearBlocks := totalTicks / This ticksPerFourYears
+		year := 4 * fourYearBlocks
+		ticksLeft := totalTicks - fourYearBlocks * This ticksPerFourYears
+		for (y in year + 1 .. year + 5) {
+			t := This ticksInYear(y)
+			if (ticksLeft < t) {
+				year = y
+				break
+			} else
+				ticksLeft -= t
+		}
+		month := 0
+		for (m in 1 .. 13) {
+			t := This ticksInMonth(year, m)
+			if (ticksLeft < t) {
+				month = m
+				break
+			} else
+				ticksLeft -= t
+		}
+		days := ticksLeft / This ticksPerDay
+		ticksLeft -= days * This ticksPerDay
+		hour := ticksLeft / This ticksPerHour
+		ticksLeft -= hour * This ticksPerHour
+		minute := ticksLeft / This ticksPerMinute
+		ticksLeft -= minute * This ticksPerMinute
+		second := ticksLeft / This ticksPerSecond
+		ticksLeft -= second * This ticksPerSecond
+		millisecond := ticksLeft / This ticksPerMillisecond
+		DateTimeData new(year, month, days + 1, hour, minute, second, millisecond)
+	}
+
+	/* returns number of ticks for given hours, minutes and seconds */
+	timeToTicks: static func (hours, minutes, seconds, millisecond: Int) -> Int64 {
+		(hours * 3600 + minutes * 60 + seconds) * This ticksPerSecond + millisecond * This ticksPerMillisecond
+	}
+
+	/* returns number of ticks for given date at 0:00*/
+	dateToTicks: static func(year, month, day : Int) -> UInt64 {
+		if (This dateIsValid(year, month, day)) {
+			totalDays := day - 1
+			for (m in 1 .. month) {
+				totalDays += This daysInMonth(year, m)
+			}
+			fourYearBlocks := (year - 1) / 4
+			year_start := fourYearBlocks * 4
+			for (y in year_start + 1 .. year)
+				totalDays += This daysInYear(y)
+			totalDays += fourYearBlocks * This daysPerFourYears
+			totalDays * This ticksPerDay
+		} else {
+			0
+		}
+	}
+
+	daysInYear: static func (year: Int) -> Int {
+		This daysPerYear + isLeapYear(year)
+	}
+
+	ticksInYear: static func (year: Int) -> UInt64 {
+		This daysInYear(year) * This ticksPerDay
+	}
+
+	daysInMonth: static func (year, month: Int) -> Int {
+		if (month == 2)
+			isLeapYear(year) ? 29 : 28
+		else if (month <= 7)
+			month % 2 ? 31 : 30
+		else
+			month % 2 ? 30 : 31
+	}
+
+	ticksInMonth: static func (year, month: Int) -> UInt64 {
+		This ticksPerDay * This daysInMonth(year, month)
+	}
+	/* validate argument ranges for hour/minutes/seconds vaules */
+	timeIsValid: static func (hour, minute, second, millisecond: Int) -> Bool {
+		hour >= 0 && hour < 24 && minute >= 0 && minute < 60 && second >= 0 && second < 60 && millisecond >= 0 && millisecond < 1000
+	}
+	/* validate argument ranges for year/month/day values */
+	dateIsValid: static func (year, month, day: Int) -> Bool {
+		year >= 1 && month >= 1 && month <= 12 && day >= 1 && day <= daysInMonth(year, month)
```

Comment:
btw. what about spacing convention with the 'in?' operator ?
should it be "month in?(1..12)" or "month in? (1 .. 12)" (is 'double dot' considered an operator ? because i used it in the 'for' loops with spaces between double-dot and the operands and now Iäm just a bit confused)